tests/test_proxy_use.py

```diff
@@ -0,0 +1,366 @@
+    with the target server is an HTTPS connection. The proxy itself will act as
+    a TCP proxy via HTTP CONNECT.
+    """
+    self.set_env_value('HTTP_PROXY', self.http_proxy_addr) # http as intended
+    self.set_env_value('HTTPS_PROXY', self.http_proxy_addr) # http as intended
+
+    self.set_env_value('REQUESTS_CA_BUNDLE',
+        os.path.join('ssl_certs', 'ssl_cert.crt'))
+
+    logger.info('Trying httpS download via http proxy: ' + self.url_https)
+    download.safe_download(self.url_https, self.target_data_length)
+    download.unsafe_download(self.url_https, self.target_data_length)
+
+
+
+
+
+  def test_http_dl_via_httpS_proxy(self):
+    """
+    Test a length-validating TUF download of a file through a proxy. Use an
+    HTTPS proxy, and perform an HTTP connection with the final server.
+    """
+    self.set_env_value('HTTP_PROXY', self.https_proxy_addr)
+    self.set_env_value('HTTPS_PROXY', self.https_proxy_addr) # unnecessary
+
+    # We're making an HTTPS connection with the proxy. The proxy will make a
+    # plain HTTP connection to the target server.
+    self.set_env_value('REQUESTS_CA_BUNDLE',
+        os.path.join('ssl_certs', 'proxy_ca.crt'))
+
+    logger.info('Trying http download via httpS proxy: ' + self.url_https)
+    download.safe_download(self.url, self.target_data_length)
+    download.unsafe_download(self.url, self.target_data_length)
+
+
+
+
+
+  def test_httpS_dl_via_httpS_proxy(self):
+    """
+    Test a length-validating TUF download of a file through a proxy. Use an
+    HTTPS proxy, and perform an HTTPS connection with the final server.
+    """
+    self.set_env_value('HTTP_PROXY', self.https_proxy_addr) # unnecessary
+    self.set_env_value('HTTPS_PROXY', self.https_proxy_addr)
+
+    # We're making an HTTPS connection with the proxy. The proxy will make its
+    # own HTTPS connection with the target server, and will have to know what
+    # certificate to trust. It was told what certs to trust when it was
+    # started in setUpClass().
+    self.set_env_value('REQUESTS_CA_BUNDLE',
+        os.path.join('ssl_certs', 'proxy_ca.crt'))
+
+    logger.info('Trying httpS download via httpS proxy: ' + self.url_https)
+    download.safe_download(self.url_https, self.target_data_length)
+    download.unsafe_download(self.url_https, self.target_data_length)
+
+
+
+
+
+  def set_env_value(self, key, value):
+    """
+    Set an environment variable after noting what the original value was, if it
+    was set, and add it to the queue for restoring to its original value / lack
+    of a value after the test finishes.
+
+    Safe for multiple uses in one test: does not overwrite original saved value
+    with new saved values.
+    """
+
+    if key in self.old_env_values:
+      # Do not save the current value. We already saved an older value, and
+      # the original one is the one we'll restore to, not whatever we most
+      # recently overwrote it with.
+      pass
+
+    elif key in os.environ:
```

Comment:
Perhaps remove the `elif`/`else` and do `self.old_env_values[key] = os.environ.get(key, None)`.

Reply:
(BTW, what about `os.execve`ing things and supplying a modified environment directly, instead of saving/restoring the current one?)

Reply:
Your suggestion reduced 6 lines to 2. Thanks!
As for execve: seems like a messy option. I'm not sure I'd want to fully replace the environment variables (instead of tweaking them) and I'm not sure I want to close the existing process and transition to a new one (temp files?). Popen execve? Not sure. I think this is probably easier and more acceptable.